### PR TITLE
Dev/mac vendors

### PR DESCRIPTION
**Pull request by ayman-m**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: Issue #22131

## Description
MACVendors.com, a free REST API to query the vendor details of a MAC address

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Integrations/MACVendors/MACVendors.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+category: Endpoint
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.yml]: i think `Utils` category fit here better

File: Integrations/MACVendors/MACVendors.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+INTEGRATION_NAME = "MAC Vendors"
+
+'''API Client'''
+
+
+class Client:
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: the `Client` should inherit from `BaseClient`

File: Integrations/MACVendors/MACVendors.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    def __init__(self, url, proxies):
+        self.base_url = url
+        self.proxies = proxies
+
+    def query(self, address=None):
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: use `BaseClient._http_request()` instead

File: Integrations/MACVendors/MACVendors.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        LOG(e)
+        return_error(e)
+    return 'ok'
+
+
+def get_mac_vendor(client):
+    args = demisto.args()
+    mac_address = args.get('address')
+    try:
+        title = ("%s - Results for MAC Address Query" % INTEGRATION_NAME)
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: use f-string instead

File: Integrations/MACVendors/MACVendors.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        raws = []
+        macvendors_ec = []
+
+        raw_response = client.query(address=mac_address)
+
+        if raw_response:
+            raws.append(raw_response)
+            macvendors_ec.append({
+                'Mac': mac_address,
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]:
```suggestion
                'MACAddress': mac_address,
```

File: Integrations/MACVendors/MACVendors.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                'Vendor': raw_response['result'].get('company'),
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: please avoid direct key access to `result`, use `.get()` instead

File: Integrations/MACVendors/MACVendors.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                'Type': raw_response['result'].get('type'),
+                'Address': raw_response['result'].get('address')
+            })
+
+        if not raws:
+            return ("%s - Could not find any results for given query" % INTEGRATION_NAME)
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: use f-string instead

File: Integrations/MACVendors/MACVendors.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        mac_address = '00:00:00'
+        client.query(address=mac_address)
+    except Exception as e:
+        LOG(e)
+        return_error(e)
+    return 'ok'
+
+
+def get_mac_vendor(client):
+    args = demisto.args()
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: pass `demisto.args()` as argument to the function

File: Integrations/MACVendors/MACVendors.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        command = demisto.command()
+        LOG('Command being called is %s' % command)
+
+        if command in commands:
+            if command == 'test-module':
+                return_outputs(test_module(client))
+
+            else:
+                outputs = commands[command](client)
+                return_outputs(outputs[0], outputs[1], outputs[2])
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: return_outputs(*commands[command](client, demisto.args()))

File: Integrations/MACVendors/MACVendors.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    try:
+        client = Client(url, proxies)
+        commands = {
+            'mac': get_mac_vendor,
+            'test-module': test_module
+        }
+
+        command = demisto.command()
+        LOG('Command being called is %s' % command)
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: use f-string

File: Integrations/MACVendors/MACVendors.yml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  id: MACVendors
+  version: -1
+configuration:
+- defaultvalue: http://macvendors.co
+  display: Server URL (e.g. https://example.net)
+  name: url
+  required: true
+  type: 0
+- display: Trust any certificate (not secure)
+  name: insecure
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.yml]: this parameter is not implemented in the code

File: Integrations/MACVendors/MACVendors.yml [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      isArray: false
+      name: address
+      required: true
+      secret: false
+    deprecated: false
+    description: MAC Address to Query Vendor For
+    execution: false
+    name: mac
+    outputs:
+    - contextPath: MACVendors.Vendor
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.yml]: please add all context outputs here

File: Integrations/MACVendors/MACVendors_description.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# Integration Instructions
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors_description.md]: this should be in `README.md` only
here you can keep the overview if you would like

File: Integrations/MACVendors/MACVendors - Test.yml [hunk 14]
```diff
@@ -0,0 +1,229 @@
+id: MACVendors - Test
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors - Test.yml]: this file should be in `TestPlaybooks` directory and should have `playbook-` prefix

File: Integrations/MACVendors/MACVendors.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        return_error(e)
+
+
+def main():
+    params = demisto.params()
+
+    # Remove trailing slash to prevent wrong URL path to service
+    url = params['url'][:-1] if (params['url'] and params['url'].endswith('/')) else params['url']
+    # Remove proxy if not set to true in params
+    proxies = handle_proxy()
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.py]: `BaseClient` call `handle_proxy`, you should just pass it the boolean values of the parameter

File: Integrations/MACVendors/MACVendors.yml [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    - default: false
+      description: The MAC Address Value
+      isArray: false
+      name: address
+      required: true
+      secret: false
+    deprecated: false
+    description: MAC Address to Query Vendor For
+    execution: false
+    name: mac
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.yml]:
```suggestion
    name: macvendors-lookup
```

File: Integrations/MACVendors/MACVendors.yml [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+description: Returns the Vendor Name of a MAC Address
+display: MAC Vendors
+name: MACVendors
+script:
+  commands:
+  - arguments:
+    - default: false
+      description: The MAC Address Value
+      isArray: false
+      name: address
```
Comment (Itay4) [inline Integrations/MACVendors/MACVendors.yml]:
```suggestion
      name: macaddress
```

**Comment by kirbles19** [comment]
@Itay4 Done.

**Comment by ayman-m** [comment]
This service's API has been updated couple of times since I submitted this PR, each API update comes with a new endpoint and new data structure, which requires a new code commit, so I will hold committing new code to the point this API stabilize and have consistent update cycles

**Comment by anara123** [comment]
@ayman-m do you recommend us to merge your PR?
Is the existing integration will still work if we will merge?

**Pull request closed without merging**